src/librustc_codegen_ssa/mir/analyze.rs

```diff
@@ -261,9 +131,67 @@ impl<'mir, 'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> Visitor<'tcx>
         self.super_terminator(terminator, location);
     }
 
-    fn visit_place(&mut self, place: &mir::Place<'tcx>, context: PlaceContext, location: Location) {
+    fn visit_place(
+        &mut self,
+        place: &mir::Place<'tcx>,
+        mut context: PlaceContext,
+        location: Location,
+    ) {
         debug!("visit_place(place={:?}, context={:?})", place, context);
-        self.process_place(&place.as_ref(), context, location);
+
+        // Non-uses do not force locals onto the stack.
+        if !context.is_use() {
+            return;
+        }
+
+        let mir::Place { local, projection } = *place;
+
+        // Reads from ZSTs do not require memory accesses and do not count when determining what
+        // needs to live on the stack.
+        if is_consume(context) {
+            let ty = place.ty(self.fx.mir, self.fx.cx.tcx()).ty;
+            let ty = self.fx.monomorphize(&ty);
+            let span = self.fx.mir.local_decls[local].source_info.span;
+            if self.fx.cx.spanned_layout_of(ty, span).is_zst() {
+                return;
+            }
+        }
+
+        let mut has_disqualifying_projection = false;
+
+        let mut projection: &[_] = projection.as_ref();
+        while let [ref proj_base @ .., elem] = *projection {
+            projection = proj_base;
+            self.super_projection_elem(local, proj_base, elem, context, location);
+
+            // Projections like `(*x)[12]` are allowed but not `*(x[12])`, since a `Deref` of a
+            // local acts like a `Copy` of that local.
```

Comment:
This is a bit iffy in that it doesn't mention *pointers* at all, which is what is being "copied", and the arbitrary example of indexing had me confused for a while, although I see what it's saying now.

I would frankly just say that a `Deref` is a pointer/reference `Copy` and be done with it.